cache/pmdk_cache.h

```diff
@@ -0,0 +1,315 @@
+
+  // TODO: set era to be current era.
+  size_t era = 0;
+  // transient fields, validated with era number:
+  TransientHandle* trans_handle = nullptr;
+
+  void Free(){
+    // TODO:
+    // free key, val, trans_handle, the transient "coat" (Block or BlockContent) of val, and this.
+  }
+  // void Ref(){
+  //   assert(trans_handle);
+  //   trans_handle->Ref();
+  // }
+  bool InCache(){
+    return in_cache;
+  }
+  bool HasRefs(){
+    if (trans_handle == nullptr){
+      return false;
+    } else {
+      return trans_handle->HasRefs();
+    }
+  }
+  void SetInCache(bool x){
+    in_cache = x;
+  }
+};
+
+using PHashTableType = po::concurrent_hash_map<po::p<uint32_t>, po::persistent_ptr<PersistentEntry>>;
+class PersistTierHashTable{
+  po::persistent_ptr<PHashTableType> table_;
+  bool KeyEqual(const char* data, size_t size, po::persistent_ptr<PersistentEntry> entry){
+    if (size != entry->key_size){
+      return false;
+    }
+    return (memcmp(data, entry->key.get(), size) == 0);
+  }
+  po::persistent_ptr<PersistentEntry>* FindPointer(const Slice& key, uint32_t hash){
+    po::persistent_ptr<PersistentEntry>* ptr;
+    PHashTableType::accessor acc;
+    bool res = table_->find(acc, hash);
+    if (res){
+      ptr = &acc->second;
+      while(*ptr != nullptr && (!KeyEqual(key.data(), key.size(), (*ptr)))) {
+        ptr = &(*ptr)->next_hash;
+      }
+    } else {
+      table_->insert(acc, PHashTableType::value_type(hash, nullptr));
+      ptr = &acc->second;
+    }
+    return ptr;
+  }
+public:
+  PersistTierHashTable(po::persistent_ptr<PHashTableType> table) : table_(table) {}
+
+  po::persistent_ptr<PersistentEntry> Insert(
+      uint32_t hash, const Slice& key, po::persistent_ptr<PersistentEntry> entry){
+    po::persistent_ptr<PersistentEntry>* ptr = FindPointer(key, hash);
+    po::persistent_ptr<PersistentEntry> old = *ptr;
+    entry->next_hash = (old == nullptr ? nullptr : old->next_hash);
+    *ptr = entry;
+    return old;
+  }
+  po::persistent_ptr<PersistentEntry> Lookup(const Slice& key, uint32_t hash){
+    return (*FindPointer(key, hash));
+  }
+  
+  po::persistent_ptr<PersistentEntry> Remove(const Slice& key, uint32_t hash){
+    // TODO
+    // this may only be used by Erase().
+  }
+  po::persistent_ptr<PersistentEntry> Remove(po::persistent_ptr<PersistentEntry> e){
+    // TODO
+    // this is used by EvictFromLRU(). consider having hash persisted in PersistentEntry.
+  }
+};
+
+struct PersistentRoot{
+  po::persistent_ptr<PHashTableType> persistent_hashtable;
+  po::persistent_ptr<PersistentEntry> persistent_lru_list;
+};
+
+// A single shard of sharded cache.
+class ALIGN_AS(CACHE_LINE_SIZE) PMDKCacheShard final : public CacheShard {
+ public:
+  PMDKCacheShard(size_t capacity, bool strict_capacity_limit,
+                double high_pri_pool_ratio, bool use_adaptive_mutex,
+                CacheMetadataChargePolicy metadata_charge_policy);
+  virtual ~PMDKCacheShard() override = default;
```

Comment:
Nit: if `override` is present, remove `virtual`. We started enforcing this after original `LRUCache` code.

Reply:
OK.